sdv/metadata/single_table.py

```diff
@@ -247,11 +246,96 @@ def to_dict(self):
+
+            unique_values = data.nunique()
+            unique_lt_categorical_threshold = unique_values <= round(len(data) / 10)
+
+            if whole_values and positive_values and unique_lt_categorical_threshold:
+                sdtype = 'categorical'
+            elif unique_values == len(data) and whole_values:
+                sdtype = 'id'
+
+        return sdtype
+
+    def _determine_sdtype_for_objects(self, data):
+        """Determine the sdtype for an object column.
+
+        Args:
+            data (pandas.Series):
+                The data to be analyzed.
+        """
+        sdtype = None
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore', category=UserWarning)
+
+                data_test = data.sample(10000) if len(data) > 10000 else data
+                datetime_format = get_datetime_format(data_test)
+                pd.to_datetime(data_test, format=datetime_format, errors='raise')
+
+            sdtype = 'datetime'
+
+        except Exception:
+            if len(data) <= 5:
+                sdtype = 'categorical'
+            else:
+                unique_values = data.nunique()
+                if unique_values == len(data):
+                    sdtype = 'id'
+                elif unique_values <= round(len(data) / 5):
+                    sdtype = 'categorical'
+                else:
+                    sdtype = 'unknown'
+
+        return sdtype
+
     def _detect_columns(self, data):
+        """Detect columns sdtype in the data.
+
+        Args:
+            data (pandas.DataFrame):
+                The data to be analyzed.
+        """
         for field in data:
-            clean_data = data[field].dropna()
-            kind = clean_data.infer_objects().dtype.kind
-            self.columns[field] = {'sdtype': self._DTYPES_TO_SDTYPES.get(kind, 'categorical')}
+            column_data = data[field]
+            clean_data = column_data.dropna()
+            dtype = clean_data.infer_objects().dtype.kind
+
+            sdtype = None
+            if dtype in self._DTYPES_TO_SDTYPES:
+                sdtype = self._DTYPES_TO_SDTYPES[dtype]
+            elif dtype in ['i', 'f']:
+                sdtype = self._determine_sdtype_for_numbers(column_data)
+
+            elif dtype == 'O':
+                sdtype = self._determine_sdtype_for_objects(column_data)
+
+            if sdtype is None:
+                raise InvalidMetadataError(
+                    f"Unsupported data type for column '{field}' (kind: {dtype})."
+                    "The valid data types are: 'object', 'int', 'float', 'datetime', 'bool'."
+                )
+            column_dict = {'sdtype': deepcopy(sdtype)}
+
+            if sdtype == 'unknown':
+                column_dict['pii'] = True
+            elif sdtype == 'datetime':
+                datetime_format = get_datetime_format(clean_data)
+                column_dict['datetime_format'] = datetime_format
```

Comment:
I think we said we'd only set the datetime format in the case where we parsed it from the text. In other words
- If the dtype is datetime
    - make the `sdtype` datetime and the `datetime_format` None
- If the dtype is object
    - if it can be converted to a datetime set the `sdtype` to datetime and set the `datetime_format`

You can maybe have the method that parses the datetime return the format if it finds it or something like that

Reply:
alright I see thanks, done in [58e4fcf](https://github.com/sdv-dev/SDV/pull/1529/commits/58e4fcf49eff5120b51b460bd92b594371086c43)